support/k8s/k8s-workload-registrar/mode-crd/controllers/pod_controller_test.go

```diff
@@ -142,6 +148,212 @@ func (s *PodControllerTestSuite) TestPodLabel() {
+			spiffeIDCount:         1,
+		},
+
+		// this section is testing identity template formatting:
+		{
+			identityTemplate: DefaultTemplate,
+			expectedSVID:     fmt.Sprintf("ns/%s/sa/%s", PodNamespace, PodServiceAccount),
+			spiffeIDCount:    1,
+		},
+		{
+			identityTemplate: "ns/{{.Pod.Namespace}}/sa/{{.Pod.ServiceAccount}}",
+			expectedSVID:     "ns/" + PodNamespace + "/sa/" + PodServiceAccount,
+			spiffeIDCount:    1,
+		},
+		{
+			identityTemplate: DefaultTemplate + "/podName/{{.Pod.Name}}",
+			expectedSVID:     "ns/" + PodNamespace + "/sa/" + PodServiceAccount + "/podName/" + PodName,
+			spiffeIDCount:    1,
+		},
+		{
+			// testing Context:
+			context: map[string]string{
+				"Region":      "EU-DE",
+				"ClusterName": "MYCLUSTER",
+			},
+			identityTemplate: "region/{{.Context.Region}}/cluster/{{.Context.ClusterName}}/podName/{{.Pod.Name}}",
+			expectedSVID:     "region/EU-DE/cluster/MYCLUSTER/podName/" + PodName,
+			spiffeIDCount:    1,
+		},
+		{
+			// testing various Pod elements:
+			identityTemplate: fmt.Sprintf("{{.Pod.%s}}/{{.Pod.%s}}/{{.Pod.%s}}/{{.Pod.%s}}/{{.Pod.%s}}", PodNameIDLabel, NamespaceIDLabel, PodServiceAccountIDLabel, PodHostnameLabel, PodNodeNameLabel),
+			expectedSVID:     PodName + "/" + PodNamespace + "/" + PodServiceAccount + "/hostname/test-node",
+			spiffeIDCount:    1,
+		},
+	}
+
+	for _, test := range tests {
+		p, err := NewPodReconciler(PodReconcilerConfig{
+			Client:                s.k8sClient,
+			Cluster:               s.cluster,
+			Ctx:                   s.ctx,
+			Log:                   s.log,
+			Scheme:                s.scheme,
+			TrustDomain:           s.trustDomain,
+			IdentityTemplate:      test.identityTemplate,
+			Context:               test.context,
+			IdentityTemplateLabel: test.identityTemplateLabel,
+		})
+		s.Require().NoError(err)
+
+		pod := corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      PodName,
+				Namespace: PodNamespace,
+				Labels:    map[string]string{test.identityTemplateLabel: test.labelValue},
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{{
+					Name:  "test-pod",
+					Image: "test-pod",
+				}},
+				NodeName:           "test-node",
+				Hostname:           "hostname",
+				ServiceAccountName: PodServiceAccount,
+			},
+		}
+		err = s.k8sClient.Create(s.ctx, &pod)
+		s.Require().NoError(err)
+		req := ctrl.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      PodName,
+				Namespace: PodNamespace,
+			},
+		}
+		_, err = p.Reconcile(req)
+		if err != nil && test.err != "" {
+			s.Require().Contains(err.Error(), test.err)
+			err = s.k8sClient.Delete(s.ctx, &pod)
+			s.Require().NoError(err)
+			continue
+		}
+		s.Require().NoError(err)
+
+		_, err = s.r.Reconcile(req)
+		if err != nil && test.err != "" {
+			s.Require().Error(err)
+			s.Require().Contains(err.Error(), test.err)
+			err = s.k8sClient.Delete(s.ctx, &pod)
+			s.Require().NoError(err)
+			continue
+		}
+		s.Require().NoError(err)
```

Comment:
This looks a little funky to me, why are we doing essentially the same call and assertions twice in a row? Are we relying on some timing of the fake K8s API server? If so, how do we know that two executions is the magic number of times we need to call `s.r.Reconcile()`?

I am also not sure about the condition `if err != nil && test.err != ""`. What happens if `test.err != "" && err == nil`? Wouldn't we want to fail the test since we expected an error but didn't get one?

Reply:
So, there are two, different issues here. 
One is the "duplicate" reconcile calls. This is based on the original code from here: https://github.com/spiffe/spire/blob/main/support/k8s/k8s-workload-registrar/mode-crd/controllers/pod_controller_test.go#L153-L157
I was initially puzzled as well, until I realized there are two different reconcile calls, for different objects, doing two different things:
* p.Reconcile()
* s.r.Reconcile()
so were should be good here.

The second part of your comment, regarding the `test.err`. The expected error can happen in several places of the flow depending on the nature of the error, and there is no easy way to indicate the exact location where the error should happen, without rewriting the entire testing logic, and that would create unnecessary complexity here. 

If you still believe this should be re-written, I can will go over this test flow in detail.

Reply:
Oh, thanks for pointing out the difference in `Reconcile()` calls, I missed that detail. One instance where the Go convention of terse variable names can bite you 😄 

> The expected error can happen in several places of the flow depending on the nature of the error, and there is no easy way to indicate the exact location where the error should happen

Would you mind elaborating on this a bit? Are the errors not deterministically returned by the same method from test execution to test execution?

Assuming the errors returned by each method are deterministic, a common way to handle this would be to have separate test template variables for errors coming from different methods, e.g. `expectPodReconcileErr`, `expectSpiffeIDReconcileErr`, etc. It looks like there are only four error cases being tested, so I would be curious to hear if there are any other challenges with adopting an approach like that.

Reply:
I was able to simplify the pod_controller test after applying the new test that you suggested. No need to create different error types anymore.